### ISSUE TITLE
Kubernetes Control Plane apiserver never reaches consensus between enlisted endpoints and num_masters config setting

### ISSUE THREAD

**Issue by lazypower** [coreos/tectonic-forum#189]
### Tectonic Version 
v1.6.8-tectonic.1

### Environment
AWS

What hardware/cloud provider/hypervisor is being used with Tectonic? AWS / t2.medium (master nodes)

### Expected Behavior

Apiserver pods should update the endpoint list without clobbering previous entries, causing false positives and weird 'round robin' routing via incorrect endpoint assignment.

### Reproduction Steps

  1. Deploy tectonic using the terraform for aws sets eg: `terraform apply`
  2. Once the cluster spin-up has completed start monitoring the controlplane elements
  3. Kick/delete one of the apiserver pods to trigger master election between the number of apiserver pods. (or just halt a master unit allowing the ASG to recreate)
  4. Observe scrolling logs near the following where the apiserver endpoint is getting updated but has nil pointers in the list (so the master_count never reaches the diff of ip's in EP and num_count of master_count)

Listing of pods:  (`kubectl get po -n kube-system -l k8s-app=kube-apiserver`)
```
$ kubectl get po -n kube-system -l k8s-app=kube-apiserver -o wide
NAME                   READY     STATUS    RESTARTS   AGE       IP             NODE
kube-apiserver-7mjqd   1/1       Running   0          7d        10.36.17.12    ip-10-36-17-12.us-west-1.compute.internal
kube-apiserver-qx9lh   1/1       Running   0          12m       10.36.45.224   ip-10-36-45-224.us-west-1.compute.internal
kube-apiserver-z9wcg   1/1       Running   1          7d        10.36.52.196   ip-10-36-52-196.us-west-1.compute.internal
```

Log output segment from kube-apiserver (`kubetail -n kube-system -l k8s-app=kube-apiserver`)
```
[kube-apiserver-qx9lh] W0907 18:30:33.364484       5 controller.go:386] Resetting endpoints for master service "kubernetes" to &{{ } {kubernetes  default /api/v1/namespaces/default/endpoints/kubernetes c10e6d9b-8dc3-11e7-8266-02c0d0e35e30 16136346 0 2017-08-30 20:42:35.517684959 +0000 UTC <nil> <nil> map[] map[] [] [] } [{[{10.36.45.224  <nil> <nil>}] [] [{https 443 TCP}]}]}
[kube-apiserver-z9wcg] W0907 18:30:39.195234       5 controller.go:386] Resetting endpoints for master service "kubernetes" to &{{ } {kubernetes  default /api/v1/namespaces/default/endpoints/kubernetes c10e6d9b-8dc3-11e7-8266-02c0d0e35e30 16136433 0 2017-08-30 20:42:35.517684959 +0000 UTC <nil> <nil> map[] map[] [] [] } [{[{10.36.52.196  <nil> <nil>}] [] [{https 443 TCP}]}]}
[kube-apiserver-7mjqd] W0907 18:30:39.756276       7 controller.go:386] Resetting endpoints for master service "kubernetes" to &{{ } {kubernetes  default /api/v1/namespaces/default/endpoints/kubernetes c10e6d9b-8dc3-11e7-8266-02c0d0e35e30 16136573 0 2017-08-30 20:42:35.517684959 +0000 UTC <nil> <nil> map[] map[] [] [] } [{[{10.36.17.12  <nil> <nil>}] [] [{https 443 TCP}]}]}
[kube-apiserver-qx9lh] W0907 18:30:43.373751       5 controller.go:386] Resetting endpoints for master service "kubernetes" to &{{ } {kubernetes  default /api/v1/namespaces/default/endpoints/kubernetes c10e6d9b-8dc3-11e7-8266-02c0d0e35e30 16136587 0 2017-08-30 20:42:35.517684959 +0000 UTC <nil> <nil> map[] map[] [] [] } [{[{10.36.45.224  <nil> <nil>}] [] [{https 443 TCP}]}]}
```

Notice the nil's in the update map?  This has some implications and i believe is triggering false positives in alert manager resulting in pages for responsive services.

This looks strikingly similar to: https://github.com/kubernetes/kubernetes/issues/38082

as well as an openshift issue here: https://github.com/openshift/origin/issues/6073

**Comment by chuckbutler** [comment]
Update:

Following along with the projected `kops` fix of just editing the endpoint and it normalized out, it seems like the apiservers dont really care what was there first and are instead overwriting with their own endpoint data regardless


```
apiVersion: v1
kind: Endpoints
metadata:
  name: kubernetes
  namespace: default
subsets:
- addresses:
  - ip: 10.36.45.224
  - ip: 10.36.17.12
  - ip: 10.36.52.196
  ports:
  - name: https
    port: 443
    protocol: TCP
```

was applied, and the apiserver is continuing to spam updates:

```
[kube-apiserver-qx9lh] W0907 18:50:27.854009       5 controller.go:386] Resetting endpoints for master service "kubernetes" to &{{ } {kubernetes  default /api/v1/namespaces/default/endpoints/kubernetes c10e6d9b-8dc3-11e7-8266-02c0d0e35e30 16165039 0 2017-08-30 20:42:35.517684959 +0000 UTC <nil> <nil> map[] map[kubectl.kubernetes.io/last-applied-configuration:{"apiVersion":"v1","kind":"Endpoints","metadata":{"annotations":{},"name":"kubernetes","namespace":"default"},"subsets":[{"addresses":[{"ip":"10.36.45.224"},{"ip":"10.36.17.12"},{"ip":"10.36.52.196"}],"ports":[{"name":"https","port":443,"protocol":"TCP"}]}]}
[kube-apiserver-qx9lh] ] [] [] } [{[{10.36.45.224  <nil> <nil>}] [] [{https 443 TCP}]}]}
```

Notice that it still has nil's in the endpoint listing, despite the prior annotation having the full ip address listing.

**Comment by chuckbutler** [comment]
Also should this be filed against the upstream kubernetes project, the tectonic-installer, or here? I had some confusion on where things like this should go overall as it looks config related, not necessarily a bug in the apiserver code.

**Comment by robszumski** [comment]
This is ultimately a Kubernetes problem, but we have the CoreOS upstream team working on a fix as we speak. I believe a proposal has already been accepted upstream. Will update here when I confirm that.

**Comment by chuckbutler** [comment]
This behavior did persist into the 1.7.3-tectonic.1 release. I can confirm as of today, its existed since 1.6.4 forward and possibly before.

**Comment by pilsy** [comment]
This is also happening on a cluster launched onto bare metal

**Comment by sym3tri** [comment]
I believe this was resolved upstream in https://github.com/kubernetes/kubernetes/pull/51698 and should be released in the 1.9 time-frame.

**Comment by Quentin-M** [comment]
Good to close.
